### PR TITLE
infowindows: none template and toggle link

### PR DESCRIPTION
**Pull request by matallo**
WIP for https://github.com/CartoDB/cartodb/issues/7418

styles for items block pending /cc @piensaenpixel 

screen shot 2016-05-18 at 12 44 01

### CODE REVIEW

**Comment by matallo** [comment]
I added the style forms in this branch, too. I feel I'm missing a lot of stuff and hardcoding some values that should belong elsewhere, but wanted to push and ask for the right direction.

the next specific issue I want to tackle would be:

- how to support "None" template, with no need to unselect all the fields (re: https://github.com/CartoDB/cartodb/issues/7418#issuecomment-220129988)
- how to show/hide form fields depending on template type
- how to support template color (time to move to compiletemplate already?)

**Comment by matallo** [comment]
CR @xavijam (for the forms)

/cc @alonsogarciapablo

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-content-view.js [hunk 1]
```diff
@@ -40,27 +50,16 @@ module.exports = cdb.core.View.extend({
...
 
-  _getModelTemplate: function () {
-    return this._layerInfowindowModel.get('template_name');
+  _resetTemplate: function () {
+    if (!this._layerInfowindowModel.get('template_name')) {
+      this._layerInfowindowModel.set('template_name', this._DEFAULT_TEMPLATE);
+    }
   },
 
   _setTemplate: function () {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-content-view.js]: Useless

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-content-view.js [hunk 2]
```diff
@@ -40,27 +50,16 @@ module.exports = cdb.core.View.extend({
     return this;
   },
 
-  _getModelTemplate: function () {
-    return this._layerInfowindowModel.get('template_name');
+  _resetTemplate: function () {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-content-view.js]: It should be in the model

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-content-view.js [hunk 3]
```diff
@@ -16,15 +19,22 @@ module.exports = cdb.core.View.extend({
...
     this._layerDefinitionModel = opts.layerDefinitionModel;
+
+    this._resetTemplate();
   },
 
   render: function () {
     this.clearSubViews();
     this.$el.empty();
 
+    var infowindow_template = _.find(this._templateStyles, function (style) {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-content-view.js]: It should be in the model (parse?)

Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-content-view.js]: `this._templateStyles` is not defined here, is it static?

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-fields-view.js [hunk 4]
```diff
@@ -24,26 +28,99 @@ module.exports = cdb.core.View.extend({
       return model.get('position');
     };
 
-    this._layerInfowindowModel.bind('change', _.debounce(function () {
+    this._setupModel();
+
+    this.add_related_model(this._layerInfowindowModel);
+  },
+
+  _setupModel: function () {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-fields-view.js]: _initBinds?

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-fields-view.js [hunk 5]
```diff
@@ -24,26 +28,99 @@ module.exports = cdb.core.View.extend({
       return model.get('position');
     };
 
-    this._layerInfowindowModel.bind('change', _.debounce(function () {
+    this._setupModel();
+
+    this.add_related_model(this._layerInfowindowModel);
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-fields-view.js]: this method is ok but it is likely it should be inside _initBinds (_setupModel) function.

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-fields-view.js [hunk 6]
```diff
@@ -24,26 +28,99 @@ module.exports = cdb.core.View.extend({
...
 
   render: function () {
     this._destroySortable();
     this.clearSubViews();
 
-    this.$el.html(template);
+    var totalFields = this._getColumnNames().length;
+    var selectedFields = this._layerInfowindowModel.fieldCount();
+
+    this.$el.html(
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-fields-view.js]: I'd check the possibility to move select/unselect all to a proper view.

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-fields-view.js [hunk 7]
```diff
@@ -24,26 +28,99 @@ module.exports = cdb.core.View.extend({
...
+    var self = this;
+
+    var names = this._sortedFields.map(function (w) { return w.get('name'); });
+    var promises = [];
+
+    _(names).each(function (f) {
+      promises.push(self._layerInfowindowModel._addField(f));
+    });
+
+    $.when.apply($, promises).then(function () {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-fields-view.js]: cc @javisantana, our promises back?

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-fields-view.js [hunk 8]
```diff
@@ -126,9 +203,8 @@ module.exports = cdb.core.View.extend({
 
     var names = this._getColumnNames();
 
-    // If there isn't any valid column available
-    // show the message in the background
-    // if (names.length == 0) {
+    // showing blank there isn't any valid column available
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-fields-view.js]: Commented code, 🔥 .

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-items-view.js [hunk 9]
```diff
@@ -12,32 +12,41 @@ module.exports = cdb.core.View.extend({
     this._layerInfowindowModel = opts.layerInfowindowModel;
     this._layerDefinitionModel = opts.layerDefinitionModel;
 
+    this._setupModel();
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-items-view.js]: _initBinds?

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-style-form-model.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
+var cdb = require('cartodb.js');
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-style-form-model.js]: infowindow-form-model not colliding with new styles views/models/etc.

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-style-form-model.js [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+var _ = require('underscore');
+var DEBOUNCE_TIME = 350;
+
+module.exports = cdb.core.Model.extend({
+
+  initialize: function (attrs, opts) {
+    if (!opts.layerInfowindowModel) throw new Error('layerInfowindowModel is required');
+    this._layerInfowindowModel = opts.layerInfowindowModel;
+    this._initBinds();
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-style-form-model.js]: First declare the schema and then bind whatever you consider.

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-style-form-model.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+      width: {
+        type: 'Number',
+        title: _t('editor.layers.infowindow.style.select-style'),
+        validators: ['required', {
+          type: 'interval',
+          min: 0,
+          max: 400
+        }]
+      },
+      headerColor: {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-style-form-model.js]: I'd not add headerColor unless infowindow type is "color".

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-style-form-view.js [hunk 13]
```diff
@@ -0,0 +1,62 @@
+var cdb = require('cartodb.js');
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-style-form-view.js]: infowindow-form-view.js

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-style-form-model.js (hunk 12 above)
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-style-form-model.js]:
```js
this.schema = { width: …};

if (whatever) {
   this.schema.color = {};
}
```

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-fields-view.js [hunk 14]
```diff
@@ -24,26 +28,99 @@ module.exports = cdb.core.View.extend({
...
     };
 
-    this._layerInfowindowModel.bind('change', _.debounce(function () {
+    this._setupModel();
+
+    this.add_related_model(this._layerInfowindowModel);
+  },
+
+  _setupModel: function () {
+    this._layerInfowindowModel.bind('change:fields', _.debounce(function () {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-fields-view.js]: Try to only save once and managed by the main view.

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-style-form-view.js [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    this.$el.html(template());
+
+    this._generateForms();
+    this._setupModel();
+
+    return this;
+  },
+
+  _setupModel: function () {
+    this._layerInfowindowModel.bind('change:fields', _.debounce(function () {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-style-form-view.js]: Commented above, not save things when model changes in children views.

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-style-form-view.js [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+      this._layerDefinitionModel.save();
+    }, 20), this);
+  },
+
+  _generateForms: function () {
+    if (this._formView) {
+      this._formView.remove();
+    }
+
+    this._formModel = new InfowindowStyleFormModel({
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-style-form-view.js]: Not style, remember.

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-style-view.js [hunk 17]
```diff
@@ -1,8 +1,9 @@
 var cdb = require('cartodb.js');
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-style-view.js]: infowindow-select | carousel-view?

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-style-view.js [hunk 18]
```diff
@@ -45,22 +47,25 @@ module.exports = cdb.core.View.extend({
...
-  //   }
-  //   this._infowindowStyleFormView = new InfowindowStyleFormView({
-  //     layerInfowindowModel: this._layerInfowindowModel
-  //   });
-  //   this.addView(this._infowindowStyleFormView);
-  //   this.$el.append(this._infowindowStyleFormView.render().el);
-  // }
+    this._infowindowStyleFormView = new InfowindowStyleFormView({
+      layerInfowindowModel: this._layerInfowindowModel,
+      min: 0,
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-style-view.js]: You don't need to set anything here, it should be in the form model directly.

**Comment by xavijam** [comment]
Miss some tests about select/unselect behaviour.

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-content-view.js (hunk 3 above)
Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-content-view.js]: I already know why I did it like that:

I use the same view for both click and hover, while having different templates
https://github.com/CartoDB/cartodb/blob/73e4b424f9abf78e41a06fe6d2f75c5fe893cd7d/lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow-view.js

shall I use a base view then extend from threre?

**Comment by matallo** [comment]
_None_ done, now working on select/unselect own view and show/hide form fields, after that I'd say it'd be ready for merge and I'll work on the header color then

**Comment by matallo** [comment]
retest this, please

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-form-model.js [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+          min: 0,
+          max: 400
+        }]
+      };
+    }
+
+    if (type === 'infowindow_light_header_blue') {
+      this.schema.headerColor = {
+        type: 'Fill',
+        // title: _t('editor.layers.infowindow.style.header-color'),
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindow-form-model.js]: this is commented for tests, how to _expect_ this?

**Comment by matallo** [comment]
I added the width in the _defaults_ to kickstart the test

however, I had previously removed the _defaults_ as infowindow and tooltip have them different i.e. https://github.com/CartoDB/cartodb.js/blob/develop/src/geo/ui/infowindow.js#L25-L36 and https://github.com/CartoDB/cartodb.js/blob/develop/src/geo/ui/tooltip.js#L7-L11 respectively) and I'm using the same infowindow-definition for them both

I'd like to discuss in further PRs how to tackle this (different definitions for each, extend the definition? please advise)

CR @javisantana 

/cc @xavijam @viddo (feedback applied)

**Comment by matallo** [comment]
finally added the last test, as @javisantana told me he can't review now, moving it to @viddo  now

**Comment by matallo** [comment]
@javisantana reviewed on site

🚀

**Pull request merged**